incydr/cli/cmds/users.py

```diff
@@ -239,33 +270,135 @@ def move(ctx: Context, user, org_guid):
+    else:  # raw-json
+        for item in roles:
+            click.echo(item.json())
+
+
 @users.command(cls=IncydrCommand)
-@click.argument("csv")
+@click.argument("file", type=AutoDecodedFile())
+@input_format_option
+@click.option("--add", "update_method", flag_value="add", default=None)
+@click.option("--remove", "update_method", flag_value="remove", default=None)
 @click.pass_context
-# TODO: role IDs
-def bulk_update_roles(ctx: Context, csv):
+def bulk_update_roles(ctx: Context, file, update_method=None, format_=None):
     """
     Bulk update roles associated with multiple users with a CSV file.
 
+     By default, the provided roles will replace the specified user's existing roles.
+     Use the --add flag or the --remove flag to add or remove roles, respectively, from a user's existing roles.
+
     Takes a single arg `CSV` which specifies the path to the file.
     Requires the following columns:
 
     * `user` - User ID or username of the user whose roles will be updated. Performs an additional lookup if username is passed.
-    * `roles` - Space-delimited list of role IDs to assign to the new user.  These will replace the specified user's existing roles.
+    * `role` - Role ID and/or role name (case-sensitive) to assign to the new user.
     """
+
+    @lru_cache()
+    def resolve_username(user):
+        if user is None:
+            return
+        elif "@" in user:
+            return user_lookup(client, user)
+        else:  # assume user_id
+            return user
+
+    class RoleUpdateCSV(CSVModel):
+        user: str = Field(csv_aliases=["user"])
+        role: str = Field(csv_aliases=["role"])
+
+    class RoleUpdateJSON(Model):
+        user: str
+        role: str
+
+    if format_ == "csv":
+        roles_file = RoleUpdateCSV.parse_csv(file)
+    else:
+        roles_file = RoleUpdateJSON.parse_json_lines(file)
+
     client = ctx.obj()
-    try:
-        for row in track(
-            read_dict_from_csv(csv),
-            description="Updated user roles...",
-            transient=True,
-        ):
-            user = row.get("user")
-            if user and "@" in user:
-                user = user_lookup(client, row["user"])
-            client.users.v1.update_roles(user, row["roles"].split(" "))
-    except HTTPError as err:
-        console.print(f"[red]Error:[/red] {err.response.text}")
+    if update_method == "add":
+
+        def update_user_roles(user_, roles):
+            client.users.v1.add_roles(user_, roles)
+
+    elif update_method == "remove":
+
+        def update_user_roles(user_, roles):
+            client.users.v1.remove_roles(user_, roles)
+
+    else:
+
+        def update_user_roles(user_, roles):
+            client.users.v1.update_roles(user_, roles)
+
+    # group updates by user
+    buckets = bucketize(
+        roles_file,
+        key=lambda role_update: role_update.user,
+        value_transform=lambda role_update: role_update.role,
+    )
+    for bucket in buckets:
```

Comment:
One potential problem with bucketing is that if there is more than one row for a user in the file, and one of those is bad but the rest are good, then they will all fail to apply.

Does the response from the server tell us which one was bad? If so, we should at least try to be specific in our error message too. If not, we should then try to re-apply them one at a time so all the valid ones get applied and we can tell the user which ones fail.

Reply:
It don't think it does, I'll double check and add the one-by-one attempts if it doesn't